### PR TITLE
Conditionally include why_participated when creating photo posts

### PR DESCRIPTION
**Pull request by aaronschachter**
#### What's this PR do?

Modifies the under-construction `topics/posts/photo` middleware to only ask user for their signup` why_participated` if it currently doesn't exist in Rogue, completing feature parity for creating photo posts via the [Gambit Campaigns `/POST campaignActivity` endpoint](https://github.com/DoSomething/gambit-campaigns/tree/master/lib/middleware/campaignActivity/photo)

* Adds request helpers for getting and saving draftSubmission values

* Modifies user signup helpers to pass specific `campaign`, `signupSource`, and `signupSourceDetails` arguments instead of a single `args` argument

* Modifies user post helpers to pass specific `campaign`, `postSource`, and post value params arguments instead of a single `args` argument

* Adds middleware comments to detail multiple question flow per https://github.com/DoSomething/gambit-conversations/pull/445/files#r235509064

* Validates input for caption and whyParticipated with `helpers.util.isValidTextFieldValue` (renaming `helpers.util.isValidTextPost`)

#### How should this be reviewed?

* Verify signups are created via keyword (delete signup if exists already to confirm)
* Verify signup is fetched via keyword if signup exists already
* Verify that you're asked for why participated when creating your 1st photo post for the signup
* Verify that you are NOT asked for why participated when creating your 2nd photo post for the signup
* Verify caption and why participated are only accepted when message text has letters and is more than 2 characters (e.g. messages with only emoji, or just a photo should return an invalid response reply)

#### Any background context you want to provide?

This should be tested on staging, as we're looking to set the `DS_GAMBIT_CONVERSATIONS_DRAFT_SUBMISSIONS_ENABLED` config to `true` to go live.

Will need to investigate how many open draft submissions there are currently before this goes live to assess whether this can go live as is (or without an import to backfill existing draft submissions into our `draftSubmissions` collection.

#### Relevant tickets

https://www.pivotaltracker.com/story/show/161494958

#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Tests added/updated for new features/bug fixes.
- [x] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

**Comment by aaronschachter** [comment]
I've set the `DS_GAMBIT_CONVERSATIONS_DRAFT_SUBMISSIONS_ENABLED` config var to `true` on staging and verified documents in the `draftSubmissions` are populated and deleted as expected as I create photo posts (testing with `friends` and `lock` keywords for photo posts).

**Comment by aaronschachter** [comment]
Querying the production campaigns DB with this query to find drafts over the past 60 days that have a photo, but were never submitted via this query (referencing [this thread](https://dba.stackexchange.com/questions/112179/date-range-query-for-past-24-hour-in-mongo-shell) for the date range)

```
db.getCollection('reportback_submissions').count({ 
  "submitted_at": { $exists: false }, 
  "photo": { $exists: true }, 
  "created_at" : { 
    $lt: new Date(), 
    $gte: new Date(new Date().setDate(new Date().getDate()-60))
  }   
})
```
Date range:

Days | Count
-----|-----
60 | 633
30 | 631
20 | 18
15 | 17

Going to dive into try to break this down into counts by campaign... but I'm leaning towards discarding these drafts for the sake of a simpler launch. Otherwise we have 2 options for getting these gambit-campaigns drafts over:

- Run a one-time script to find create new `draftSubmissions` documents in this app with the corresponding `reportback_submissions` documents in gambit-campaigns
- Create an endpoint in gambit-campaigns to lookup a `reportback_submission` by user and campaign, and check to see if a record exists before creating a new `draftSubmission` in this app

The 1st option would be less overhead, and could be something we could run after launch. Treat Yo Friends is a photo post campaign launching this weekend, so ideally looking to enable the `draftSubmission` document before the launch, as then we'd need to execute option 1 or 2 before we can enable the feature.

**Comment by aaronschachter** [comment]
Breaking down the 611 incomplete drafts that have created date within 30 days from now:

* 586 are for VSpot (https://gambit-admin.herokuapp.com/campaigns/7314)
* 19 are for Friends On Lock (https://gambit-admin.herokuapp.com/campaigns/8226)
* 4 are for Grab the Mic (https://gambit-admin.herokuapp.com/campaigns/8017)
* 2 are for Youth Wave of Action (https://gambit-admin.herokuapp.com/campaigns/8228)
* 0 are for Mirror Messages (https://gambit-admin.herokuapp.com/campaigns/7)

cc @ngjo

File: test/unit/lib/middleware/messages/member/topics/posts/photo/draft-caption.test.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+});
+
+test('draftCaption should call save draft value if request does not have draft value and has valid text field value, and call next if hasSignupWithWhyParticipated', async (t) => {
+  const next = sinon.stub();
+  const middleware = draftCaption(mockConfig);
+  sandbox.stub(helpers.topic, 'isPhotoPostConfig')
+    .returns(true);
+  sandbox.stub(helpers.request, 'hasDraftSubmissionValue')
+    .returns(false);
+  sandbox.stub(helpers.util, 'isValidTextFieldValue')
```
Comment (DFurnes) [inline test/unit/lib/middleware/messages/member/topics/posts/photo/draft-caption.test.js]: This isn't specific to this pull request, but I'm a little nervous about how thoroughly we mock things & how much of the implementation details are mirrored in these tests. It seems like it'd make it hard to refactor the app code without changing a lot of these tests.

Ideally we'd only mock external dependencies (like API calls), so we could test the full behavior of this part of the app: "given _this_ input to this part of the app, with external resources behaving in _this way_, I get _that_ result". So for this test, something like this:

```js
const next = sinon.stub();
const middleware = draftCaption(config);

// mock the database state
sandbox.stub(helpers.request, 'hasDraftSubmissionValue')
  .returns(false);
sandbox.stub(helpers.request, 'saveDraftSubmissionValue')
  .returns(Promise.resolve(true));

// test
await middleware(t.context.req, t.context.res, next);

// assertions
helpers.request.saveDraftSubmissionValue
  .should.have.been.calledWith(t.context.req, mockConfig.draftValueKey, mockInboundMessageText);
next.should.have.been.called;
helpers.replies.askWhyParticipated.should.not.have.been.called;
helpers.replies.invalidCaption.should.not.have.been.called;

```

File: config/lib/middleware/messages/member/topics/posts/photo/draft-photo.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = {
+  draftValueKey: 'url',
```
[the remark this answers is not in the document]
Comment (aaronschachter) [reply config/lib/middleware/messages/member/topics/posts/photo/draft-photo.js]: Great call; making an update for this.

**Pull request merged**